### PR TITLE
F service platform UI

### PR DESCRIPTION
**Pull request by 4eck-qed**
Depends on new localization keys.

### CODE REVIEW

File: src/Simplic.PlugIn.ServicePlatform.Client/packages.config [hunk 1]
```diff
@@ -0,0 +1,4 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="Newtonsoft.Json" version="13.0.1" targetFramework="net48" />
+</packages>
```
Comment (MEichhoff) [inline src/Simplic.PlugIn.ServicePlatform.Client/packages.config]: migrate package config. This is a context menu function in visual studio. After the migration installed packages will be referenced as nuget reference and not in a seperate file

File: src/Simplic.ServicePlatform.Data.DB/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -32,5 +32,5 @@
 // You can specify all the values or you can default the Build and Revision Numbers
 // by using the '*' as shown below:
 // [assembly: AssemblyVersion("1.0.*")]
-[assembly: AssemblyVersion("1.0.1221.1230")]
-[assembly: AssemblyFileVersion("1.0.1221.1230")]
+[assembly: AssemblyVersion("1.0.3421.1230")]
+[assembly: AssemblyFileVersion("1.0.3421.1230")]
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.Data.DB/Properties/AssemblyInfo.cs]: usually you should use the pipeline to increase the version number and not push the version of your local nuget manager

File: src/Simplic.ServicePlatform.UI/Behaviors/CopyItemDragDropBehavior.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+namespace Simplic.ServicePlatform.UI
+{
+    /// <summary>
+    /// Custom ListBoxDragDropBehavior that prevents removal of items when drag & drop is completed.
+    /// </summary>
+    public class CopyItemDragDropBehavior : ListBoxDragDropBehavior
+    {
+        public override void DragDropCompleted(DragDropState state) { }
+    }
+}
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/Behaviors/CopyItemDragDropBehavior.cs:12]: is this still required wenn it does not do anything?

File: src/Simplic.ServicePlatform.UI/Behaviors/UniqueServiceModuleDragDropBehavior.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    public class UniqueServiceModuleDragDropBehavior : ListBoxDragDropBehavior
+    {
+        public override void Drop(DragDropState state)
+        {
+            var draggedItems = state.DraggedItems as IEnumerable<ServiceModuleViewModel>;
+            var draggedModule = draggedItems?.FirstOrDefault();
+            if (state.DestinationItemsSource is IEnumerable<ServiceModuleViewModel> modules &&
+                modules.Any(m => m.Model.Name == draggedModule?.Model.Name))
+            {
+                MessageBox.Show(CommonServiceLocator.ServiceLocator.Current.GetInstance<ILocalizationService>().Translate("validation_already_exists", draggedModule?.Model.Name));
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/Behaviors/UniqueServiceModuleDragDropBehavior.cs]: use the localized message box instead.

File: src/Simplic.ServicePlatform.UI/Behaviors/UniqueServiceModuleDragDropBehavior.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+using Telerik.Windows.DragDrop.Behaviors;
+
+namespace Simplic.ServicePlatform.UI
+{
+    /// <summary>
+    /// Custom ListBoxDragDropBehavior that prevents dropping of redundant items.
+    /// </summary>
+    public class UniqueServiceModuleDragDropBehavior : ListBoxDragDropBehavior
+    {
+        public override void Drop(DragDropState state)
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/Behaviors/UniqueServiceModuleDragDropBehavior.cs:16]: all public things should have comments

File: src/Simplic.ServicePlatform.UI/Converters/ModuleDataConverter.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/Converters/ModuleDataConverter.cs]: remove unnecessary usings

File: src/Simplic.ServicePlatform.UI/Converters/ModuleDataConverter.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+using System.Threading.Tasks;
+using Telerik.Windows.DragDrop.Behaviors;
+
+namespace Simplic.ServicePlatform.UI
+{
+    public class ModuleDataConverter : DataConverter
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/Converters/ModuleDataConverter.cs:10]: comment

File: src/Simplic.ServicePlatform.UI/ViewModels/ServiceDefinitionViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+using Simplic.Localization;
+
+namespace Simplic.ServicePlatform.UI
+{
+    /// <summary>
+    /// View model for the service definition.
+    /// </summary>
+    public class ServiceDefinitionViewModel : ViewModelBase, IDataErrorInfo
+    {
+        #region Fields
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/ViewModels/ServiceDefinitionViewModel.cs]: we dont want to use regions

File: src/Simplic.ServicePlatform.UI/ViewModels/ServiceDefinitionViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+            if (string.IsNullOrWhiteSpace(ServiceName))
+                return localization.Translate("itemBoxProfile_display_name_empty");
+
+            if (Parent.Services.Count(x => string.Equals(x.ServiceName, ServiceName, StringComparison.CurrentCultureIgnoreCase)) > 1)
+                return localization.Translate("xaml_name_assigned");
+
+            if (ServiceName.Length < 3)
+                return localization.Translate("validation_name_at_least", "3");
+            return string.Empty;
+        }
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/ViewModels/ServiceDefinitionViewModel.cs]: move to private methods

File: src/Simplic.ServicePlatform.UI/ViewModels/ServiceViewModel.cs [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+using System.Windows;
+using System.Windows.Data;
+using System.Windows.Input;
+using System.Windows.Threading;
+using Simplic.Localization;
+
+namespace Simplic.ServicePlatform.UI
+{
+
+    /// <summary>
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/ViewModels/ServiceViewModel.cs]:
```suggestion
{
    /// <summary>
```

File: src/Simplic.ServicePlatform.UI/ViewModels/ServiceViewModel.cs [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+    /// View model for the service.
+    /// </summary>
+    public class ServiceViewModel : ViewModelBase
+    {
+        #region Fields
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/ViewModels/ServiceViewModel.cs]: remove regions

File: src/Simplic.ServicePlatform.UI/ViewModels/ServiceViewModel.cs [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+        #endregion
+
+        /// <summary>
+        /// Instantiates the view model.
+        /// </summary>
+        /// <param name="serviceClient">Service client</param>
+        public ServiceViewModel(IServiceClient serviceClient)
+        {
+
+            this.serviceClient = serviceClient;
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/ViewModels/ServiceViewModel.cs]:
```suggestion
        {
            this.serviceClient = serviceClient;
```

File: src/Simplic.ServicePlatform.UI/Views/ServiceView.xaml.cs [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+                Source = DataContext
+            });
+
+            cardButtonGroup.Items.Add(removeCardButton);
+
+            if (DataContext is ServiceViewModel viewModel)
+            {
+
+                addCardButton.Command = viewModel.AddCardCommand;
+            }
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/Views/ServiceView.xaml.cs]:
```suggestion
            {
                addCardButton.Command = viewModel.AddCardCommand;
            }
```

**Review by MEichhoff — CHANGES_REQUESTED** [review]
Also include a screenshot of the ui

File: src/Simplic.ServicePlatform.UI/Behaviors/CopyItemDragDropBehavior.cs (hunk 3 above)
Comment (4eck-qed) [reply src/Simplic.ServicePlatform.UI/Behaviors/CopyItemDragDropBehavior.cs:12]: yes it is supposed to do nothing in so items wont get removed on drag drop completion.

File: src/Simplic.ServicePlatform.UI/Behaviors/UniqueServiceModuleDragDropBehavior.cs (hunk 4 above)
Comment (4eck-qed) [reply src/Simplic.ServicePlatform.UI/Behaviors/UniqueServiceModuleDragDropBehavior.cs]: Add https://github.com/simplic/simplic-studio/pull/55 to the official package.

File: src/Simplic.ServicePlatform.Data.DB/Properties/AssemblyInfo.cs [hunk 14]
```diff
@@ -32,5 +32,5 @@
 // You can specify all the values or you can default the Build and Revision Numbers
 // by using the '*' as shown below:
 // [assembly: AssemblyVersion("1.0.*")]
-[assembly: AssemblyVersion("1.0.1221.1230")]
-[assembly: AssemblyFileVersion("1.0.1221.1230")]
+[assembly: AssemblyVersion("1.0.0.0")]
+[assembly: AssemblyFileVersion("1.0.0.0")]
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.Data.DB/Properties/AssemblyInfo.cs]:
```suggestion
[assembly: AssemblyVersion("1.1.22.323")]
[assembly: AssemblyFileVersion("1.1.22.323")]
```

File: src/Simplic.ServicePlatform.UI/Converters/ModuleDataConverter.cs (hunk 7 above)
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/Converters/ModuleDataConverter.cs:10]:
```suggestion
    /// <summary>
    /// Converter for modules to service definitions.
    /// </summary>
    public class ModuleDataConverter : DataConverter
```

File: src/Simplic.ServicePlatform.UI/Converters/NullToBooleanConverter.cs [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+        /// <param name="parameter"></param>
+        /// <param name="culture"></param>
+        /// <returns>Not implemented.</returns>
+        /// <exception cref="NotImplementedException"></exception>
+        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
+        {
+            throw new NotImplementedException();
+        }
+
+    }
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/Converters/NullToBooleanConverter.cs]:
```suggestion
        }
    }
```

File: src/Simplic.ServicePlatform.UI/Converters/NullToBooleanConverter.cs [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+
+}
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/Converters/NullToBooleanConverter.cs]:
```suggestion
    }
}
```

File: src/Simplic.ServicePlatform.UI/Properties/AssemblyInfo.cs [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+//      Major Version
+//      Minor Version
+//      Build Number
+//      Revision
+//
+// You can specify all the values or you can default the Build and Revision Numbers
+// by using the '*' as shown below:
+// [assembly: AssemblyVersion("1.0.*")]
+[assembly: AssemblyVersion("1.0.38400.0")]
+[assembly: AssemblyFileVersion("1.0.38400.0")]
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/Properties/AssemblyInfo.cs]:
```suggestion
[assembly: AssemblyVersion("1.0.22.323")]
[assembly: AssemblyFileVersion("1.0.22.323")]
```

File: src/Simplic.ServicePlatform.UI/ViewModels/ServiceDefinitionViewModel.cs [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+                RaisePropertyChanged(nameof(ServiceName));
+            });
+        }
+
+        private void Timer_Tick(object sender, EventArgs e)
+        {
+            RenameCommand.Execute(this);
+            timer.Stop();
+        }
+        private string CheckServiceName()
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/ViewModels/ServiceDefinitionViewModel.cs]:
```suggestion
        }
        
        private string CheckServiceName()
```

**Comment by FabianFS10** [comment]
image

File: src/Simplic.ServicePlatform.UI/ViewModels/ServiceViewModel.cs [hunk 19]
```diff
@@ -0,0 +1,273 @@
...
+
+                CheckRenameAndRegisterRemoval(service);
+            }
+
+            RemoveServices();
+            
+            if (errors) LocalizedMessageBox.Show("error_save_services", "Error", MessageBoxButton.OK, MessageBoxImage.Information);
+
+
+        }
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/ViewModels/ServiceViewModel.cs]:
```suggestion
            if (errors) LocalizedMessageBox.Show("error_save_services", "Error", MessageBoxButton.OK, MessageBoxImage.Information);
        }
```

**Review by MEichhoff — CHANGES_REQUESTED** [review]

**Comment by MEichhoff** [comment]
image

**Comment by MEichhoff** [comment]
also remove the ':' after Dienstname

**Comment by FabianFS10** [comment]
image

**Comment by MEichhoff** [comment]
the space between the name nad the config can be smaller, also please include a screenshot with an opened configuration

**Comment by FabianFS10** [comment]
image

**Comment by MEichhoff** [comment]
still the space between the name and the body can be reduced

**Comment by FabianFS10** [comment]
image

**Comment by FabianFS10** [comment]
Textbox is now set on stretch, the space was minimal reduced. It has to be like that because of localization the text box has to adjust to the label before the text box.

**Comment by FabianFS10** [comment]
image

File: src/Simplic.ServicePlatform.UI/ServicePlattformApplicationHelper.cs [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+            IServiceClient serviceClient = CommonServiceLocator.ServiceLocator.Current.GetInstance<IServiceClient>();
+            Application.Current.Dispatcher.Invoke(async () =>
+            {
+                var service = await serviceClient.GetAllServices();
+                var view = new ServiceView(serviceClient);
+                view.Show();
+            });
+
+
+        }
```
Comment (MEichhoff) [inline src/Simplic.ServicePlatform.UI/ServicePlattformApplicationHelper.cs]:
```suggestion
            });
        }
```

**Review by MEichhoff — CHANGES_REQUESTED** [review]

**Review by MEichhoff — APPROVED** [review]

**Pull request merged**